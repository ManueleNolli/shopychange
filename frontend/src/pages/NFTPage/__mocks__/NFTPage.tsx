import { divWithoutChildrenMock } from '../../../../mocksUtils'

const NFTPageMock = () => divWithoutChildrenMock('NFTPage')
export default NFTPageMock
