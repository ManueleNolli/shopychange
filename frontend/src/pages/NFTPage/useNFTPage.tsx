import { useState, useCallback, useEffect } from 'react'
import { Placeholder } from '../../assets/AssetsManager'
import { useUserContext } from '../../context/userContext'
import {
  nftCompleteQuery,
  saleQuery,
  nftHistoryQuery,
  handlerQueryErrorCatch,
} from '../../services/BackendService'
import { NFTHistory } from '../../types/components/Events'
import { NFTComplete } from '../../types/components/NFTComplete'
import { Sale } from '../../types/components/Sale'
import CheckImageSrc from '../../utils/CheckImageSrc/CheckImageSrc'
import { NFTPageProps } from '../../types/props/NFTPageProps'
import useFetching from '../../hooks/useFetching'

export default function useNFTPage({ contractAddress, tokenId }: NFTPageProps) {
  const { userAddress, blockchainNetworkId } = useUserContext()

  const [image, setImage] = useState<string>(Placeholder)

  const [nftComplete, setNftComplete] = useState<NFTComplete>({
    nft: {
      name: '',
      image: '',
      contractAddress: '',
      tokenId: 0,
    },
    owner: '',
    description: '',
    attributes: [],
    contract: {
      name: '',
      symbol: '',
      address: '',
    },
    tokenType: '',
  })
  const {
    isFetching: isNftCompleteFetching,
    setIsFetching: setIsNftCompleteFetching,
    isError: isNftCompleteError,
    setIsError: setIsNftCompleteError,
  } = useFetching()

  const [sale, setSale] = useState<Sale | null>(null)
  const {
    isFetching: isSaleFetching,
    setIsFetching: setIsSaleFetching,
    isError: isSaleError,
    setIsError: setIsSaleError,
  } = useFetching()

  const [history, setHistory] = useState<NFTHistory>(
    [] as unknown as NFTHistory
  )
  const {
    isFetching: isHistoryFetching,
    setIsFetching: setIsHistoryFetching,
    isError: isHistoryError,
    setIsError: setIsHistoryError,
  } = useFetching()

  const fetchNFTComplete = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      nftCompleteQuery(blockchainNetworkId, contractAddress, tokenId)
    )

    if (response.error) {
      setIsNftCompleteError(true)
      setIsNftCompleteFetching(false)
      return
    }

    const result = response.result
    const nftComplete = result.nftComplete as NFTComplete
    setNftComplete(nftComplete)
    setIsNftCompleteFetching(false)
    return nftComplete
  }

  const fetchSale = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      saleQuery(blockchainNetworkId, contractAddress, tokenId)
    )

    if (response.error) {
      setIsSaleError(true)
      setIsSaleFetching(false)
      return
    }

    const result = response.result

    const sale = result.sale as Sale
    setSale(sale)
    setIsSaleFetching(false)
  }

  const fetchHistory = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      nftHistoryQuery(blockchainNetworkId, contractAddress, tokenId)
    )

    if (response.error) {
      setIsHistoryError(true)
      setIsHistoryFetching(false)
      return
    }

    const result = response.result
    const history = result.nftHistory as NFTHistory
    setHistory(history)
    setIsHistoryFetching(false)
  }

  const updateAll = useCallback(async () => {
    fetchSale()
    fetchHistory()
    fetchNFTComplete()
  }, [contractAddress, tokenId])

  useEffect(() => {
    const fetchImage = async (nftCompleteFetched: NFTComplete | undefined) => {
      if (nftCompleteFetched === undefined) return
      const exists = await CheckImageSrc(nftCompleteFetched.nft.image)
      if (exists) setImage(nftCompleteFetched.nft.image)
    }
    fetchNFTComplete().then(fetchImage)
    fetchSale()
    fetchHistory()
  }, [])

  return {
    image,
    nftComplete,
    isNftCompleteFetching,
    isNftCompleteError,
    sale,
    isSaleFetching,
    isSaleError,
    history,
    isHistoryFetching,
    isHistoryError,
    updateAll,
  }
}
