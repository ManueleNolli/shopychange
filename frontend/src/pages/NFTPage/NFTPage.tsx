import React from 'react'
import {
  Box,
  Center,
  Heading,
  Text,
  Image,
  Flex,
  Divider,
  TableContainer,
  Table,
  Tbody,
  Td,
  Tr,
  SimpleGrid,
  Spacer,
} from '@chakra-ui/react'

import NFTOperations from '../../components/NFTOperations/NFTOperations'

import { NFTPageProps } from '../../types/props/NFTPageProps'

import { renderAddress } from '../../utils/SimplifiedAddress/SimplifiedAddress'

import { Attribute } from '../../types/components/Attribute'
import { NFTHistory } from '../../types/components/Events'
import NFTHistoryTable from '../../components/NFTHistoryTable/NFTHistoryTable'
import { Sale } from '../../types/components/Sale'
import { NFT } from '../../types/components/NFT'
import useNFTPage from './useNFTPage'
import QueryContainer from '../../utils/QueryContainer/QueryContainer'

export default function NFTPage({ contractAddress, tokenId }: NFTPageProps) {
  const {
    image,
    nftComplete,
    isNftCompleteFetching,
    isNftCompleteError,
    sale,
    isSaleFetching,
    isSaleError,
    history,
    isHistoryFetching,
    isHistoryError,
    updateAll,
  } = useNFTPage({ contractAddress, tokenId })

  const renderOperations = (
    sale: Sale | null,
    isSaleFetching: boolean,
    isSaleError: boolean,
    nft: NFT,
    owner: string
  ) => {
    return (
      <QueryContainer isLoading={isSaleFetching} isError={isSaleError}>
        <NFTOperations
          nft={nft}
          owner={owner}
          sale={sale}
          onUpdate={updateAll}
        />
      </QueryContainer>
    )
  }

  const renderHistory = (
    history: NFTHistory,
    isHistoryFetching: boolean,
    isHistoryError: boolean
  ) => {
    return (
      <QueryContainer isLoading={isHistoryFetching} isError={isHistoryError}>
        <NFTHistoryTable nftHistory={history} />
      </QueryContainer>
    )
  }

  const renderAttribute = (attribute: Attribute) => {
    return (
      <Flex
        minW={'8vw'}
        minH={'8vh'}
        key={attribute.traitType}
        borderColor={'gray.200'}
        borderWidth={'1px'}
        flexDir={'column'}
        rounded={'lg'}
      >
        <Spacer />
        <Center>
          <Text color={'gray.400'}>{attribute.traitType}</Text>
        </Center>
        <Spacer />
        <Center>
          <Text>{attribute.value}</Text>
        </Center>
        <Spacer />
      </Flex>
    )
  }

  return (
    <Box>
      <QueryContainer
        isLoading={isNftCompleteFetching}
        isError={isNftCompleteError}
      >
        <Center>
          <Flex direction="column">
            <Heading textAlign={'center'}>{nftComplete.nft.name}</Heading>
            <Center>
              <Box
                rounded={'lg'}
                mt={12}
                pos={'relative'}
                height={'50vh'}
                width={'50vh'}
                _after={{
                  transition: 'all .3s ease',
                  content: '""',
                  w: 'full',
                  h: 'full',
                  pos: 'absolute',
                  top: 0,
                  left: 0,
                  backgroundImage: `url(${image})`,
                  backgroundSize: 'cover',
                  filter: 'blur(25px)',
                  zIndex: -1,
                }}
                _hover={{
                  _after: {
                    filter: 'blur(35px)',
                  },
                }}
              >
                <Center mt={'5vh'}>
                  <Image
                    rounded={'lg'}
                    height="40vh"
                    width="40vh"
                    objectFit={'cover'}
                    src={image}
                  />
                </Center>
              </Box>
            </Center>

            <Box my={5}>
              {renderOperations(
                sale,
                isSaleFetching,
                isSaleError,
                nftComplete.nft,
                nftComplete.owner
              )}
            </Box>

            <Center>
              <Box>
                <Flex
                  w={{ base: '80vw', md: '70vw' }}
                  direction={{ base: 'column', xl: 'row' }}
                >
                  <Box my={5} flex="1">
                    <Heading mb={5} textAlign={'center'} size={'md'}>
                      Description
                    </Heading>
                    <Text textAlign={'center'} wordBreak={'break-word'}>
                      {nftComplete.description}
                    </Text>
                  </Box>

                  <Box m={5}>
                    <Divider orientation="vertical" size={'50px'} />
                    <Divider />
                  </Box>

                  <Box my={5} flex="1">
                    <Heading mb={5} textAlign={'center'} size={'md'}>
                      Details
                    </Heading>
                    <TableContainer>
                      <Table variant="unstyled">
                        <Tbody>
                          <Tr>
                            <Td py="2">Contract Address</Td>
                            <Td py="2" textAlign={'end'}>
                              {renderAddress(
                                nftComplete.contract.address,
                                `/${nftComplete.contract.address}`
                              )}
                            </Td>
                          </Tr>
                          <Tr>
                            <Td py="2">Contract Name</Td>
                            <Td py="2" textAlign={'end'}>
                              {nftComplete.contract.name}
                            </Td>
                          </Tr>
                          <Tr>
                            <Td py="2">Contract Symbol</Td>
                            <Td py="2" textAlign={'end'}>
                              {nftComplete.contract.symbol}
                            </Td>
                          </Tr>
                          <Tr>
                            <Td py="2">Token ID</Td>
                            <Td py="2" textAlign={'end'}>
                              {nftComplete.nft.tokenId}
                            </Td>
                          </Tr>
                          <Tr>
                            <Td py="2">Token Type</Td>
                            <Td py="2" textAlign={'end'}>
                              {nftComplete.tokenType}
                            </Td>
                          </Tr>
                          <Tr>
                            <Td py="2">Owner</Td>
                            <Td py="2" textAlign={'end'}>
                              {renderAddress(
                                nftComplete.owner,
                                `/account/${nftComplete.owner}`
                              )}
                            </Td>
                          </Tr>
                        </Tbody>
                      </Table>
                    </TableContainer>
                  </Box>
                </Flex>
                <Divider my={5} />
                {nftComplete.attributes.length > 0 && (
                  <>
                    <Box>
                      <Heading mb={5} textAlign={'center'} size={'md'}>
                        Attributes
                      </Heading>

                      <SimpleGrid
                        columns={{ sm: 2, md: 3, lg: 4, xl: 5 }}
                        spacing={10}
                      >
                        {nftComplete.attributes.map((attribute) =>
                          renderAttribute(attribute)
                        )}
                      </SimpleGrid>
                    </Box>
                    <Divider my={5} />
                  </>
                )}
                <Box>
                  <Heading mb={5} textAlign={'center'} size={'md'}>
                    History
                  </Heading>
                  {renderHistory(history, isHistoryFetching, isHistoryError)}
                </Box>
              </Box>
            </Center>
          </Flex>
        </Center>
      </QueryContainer>
    </Box>
  )
}
