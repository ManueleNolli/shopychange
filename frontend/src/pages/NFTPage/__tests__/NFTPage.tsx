import React from 'react'
import { render } from '@testing-library/react'
import NFTPage from '../NFTPage'
import useNFTPage from '../useNFTPage'

jest.mock('../../../utils/SimplifiedAddress/SimplifiedAddress', () => ({
  renderAddress: jest.fn((address: string) => address),
}))

jest.mock('../../../utils/FormatDate/FormatDate', () => ({
  formatDate: jest.fn((data: string) => data),
}))

jest.mock('../../../utils/CorrectDate/CorrectDate', () => ({
  correctDate: jest.fn((data: string) => data),
}))

jest.mock('../../../components/NFTHistoryTable/NFTHistoryTable')
jest.mock('../../../components/NFTOperations/NFTOperations')

jest.mock('../useNFTPage', () => jest.fn())
const mockProps = {
  contractAddress: 'mockedContractAddress',
  tokenId: 3,
}
const mockedData = {
  nftComplete: {
    nft: {
      contractAddress: 'mockedContractAddress',
      tokenId: 0,
      image: 'mockedImage',
      name: 'mockedName',
    },
    description: 'mockedDescription',
    contract: {
      name: 'mockedContractName',
      symbol: 'mockedContractSymbol',
      address: 'mockedContractAddress',
    },
    attributes: [
      {
        traitType: 'mockedTraitType',
        value: 'mockedValue',
      },
    ],
  },
  sale: {
    nftData: {
      contractAddress: 'mockedContractAddress',
      tokenId: 0,
      image: 'mockedImage',
      name: 'mockedName',
    },
    seller: 'mockedSeller',
    price: 0,
    status: 0,
  },
  nftHistory: [],
}
describe('NFTPage', () => {
  test('renders successfully', async () => {
    ;(useNFTPage as jest.Mock).mockReturnValue({
      image: 'mockedImage',
      nftComplete: mockedData.nftComplete,
      isNftCompleteFetching: false,
      isNftCompleteError: false,
      sale: mockedData.sale,
      isSaleFetching: false,
      isSaleError: false,
      history: mockedData.nftHistory,
      isHistoryFetching: false,
      isHistoryError: false,
      updateAll: jest.fn(),
    })
    const tree = render(<NFTPage {...mockProps} />)

    expect(tree.container).toMatchSnapshot()
  })
})
