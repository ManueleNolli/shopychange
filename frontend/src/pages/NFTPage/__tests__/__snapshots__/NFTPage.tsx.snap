// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTPage renders successfully 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-gmuwbf"
    >
      <div
        class="css-j7qwjs"
      >
        <h2
          class="chakra-heading css-xi606m"
        >
          mockedName
        </h2>
        <div
          class="css-gmuwbf"
        >
          <div
            class="css-1om72w7"
          >
            <div
              class="css-1ffdipv"
            >
              <img
                class="chakra-image css-1ofc2kw"
                src="mockedImage"
              />
            </div>
          </div>
        </div>
        <div
          class="css-s98few"
        >
          <div
            data-testid="NFTOperations"
          />
        </div>
        <div
          class="css-gmuwbf"
        >
          <div
            class="css-0"
          >
            <div
              class="css-gckdu1"
            >
              <div
                class="css-13i8xqh"
              >
                <h2
                  class="chakra-heading css-fxkt9r"
                >
                  Description
                </h2>
                <p
                  class="chakra-text css-2tly1c"
                >
                  mockedDescription
                </p>
              </div>
              <div
                class="css-1522hk"
              >
                <hr
                  aria-orientation="vertical"
                  class="chakra-divider css-12gj880"
                />
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-1upb9tn"
                />
              </div>
              <div
                class="css-13i8xqh"
              >
                <h2
                  class="chakra-heading css-fxkt9r"
                >
                  Details
                </h2>
                <div
                  class="chakra-table__container css-zipzvv"
                >
                  <table
                    class="chakra-table css-0"
                  >
                    <tbody
                      class="css-0"
                    >
                      <tr
                        class="css-0"
                      >
                        <td
                          class="css-o5g2qp"
                        >
                          Contract Address
                        </td>
                        <td
                          class="css-r7o798"
                        />
                      </tr>
                      <tr
                        class="css-0"
                      >
                        <td
                          class="css-o5g2qp"
                        >
                          Contract Name
                        </td>
                        <td
                          class="css-r7o798"
                        >
                          mockedContractName
                        </td>
                      </tr>
                      <tr
                        class="css-0"
                      >
                        <td
                          class="css-o5g2qp"
                        >
                          Contract Symbol
                        </td>
                        <td
                          class="css-r7o798"
                        >
                          mockedContractSymbol
                        </td>
                      </tr>
                      <tr
                        class="css-0"
                      >
                        <td
                          class="css-o5g2qp"
                        >
                          Token ID
                        </td>
                        <td
                          class="css-r7o798"
                        >
                          0
                        </td>
                      </tr>
                      <tr
                        class="css-0"
                      >
                        <td
                          class="css-o5g2qp"
                        >
                          Token Type
                        </td>
                        <td
                          class="css-r7o798"
                        />
                      </tr>
                      <tr
                        class="css-0"
                      >
                        <td
                          class="css-o5g2qp"
                        >
                          Owner
                        </td>
                        <td
                          class="css-r7o798"
                        />
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-10qd4hr"
            />
            <div
              class="css-0"
            >
              <h2
                class="chakra-heading css-fxkt9r"
              >
                Attributes
              </h2>
              <div
                class="css-15q3nny"
              >
                <div
                  class="css-cj6ru4"
                >
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-gmuwbf"
                  >
                    <p
                      class="chakra-text css-qljqz"
                    >
                      mockedTraitType
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-gmuwbf"
                  >
                    <p
                      class="chakra-text css-0"
                    >
                      mockedValue
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                </div>
              </div>
            </div>
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-10qd4hr"
            />
            <div
              class="css-0"
            >
              <h2
                class="chakra-heading css-fxkt9r"
              >
                History
              </h2>
              <div
                data-testid="NFTHistoryTable"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
